File: php/Response/Response.php
<?php

require_once("EasyJSON\EasyJSON.php");
use EasyJSON\JsonSerializable as JsonSerializable;

class ResponseStatus {

    const OK = "ok";
    const ERROR = "error";

}

class Response implements JsonSerializable {

    const STATUS_TAG = "status";
    const TITLE_TAG = "title";
    const MESSAGE_TAG = "message";
    const DATETIME_TAG = "datetime";
    const DATA_TAG = "data";

    private $status;
    private $title;
    private $message;
    private $date;
    private $data;

    /**
     * Response constructor
     * @param $status
     * @param $title
     * @param $message
     */
    public function __construct($status, $title, $message, $data) {
        $this->status = $status;
        $this->title = $title;
        $this->message = $message;
        $this->date = date("Y-m-d H:i:s");
        $this->data = $data;
    }

    /**
     * @return mixed
     */
    public function getStatus() {
        return $this->status;
    }

    /**
     * @param mixed $status
     */
    public function setStatus($status) {
        $this->status = $status;
    }

    /**
     * @return mixed
     */
    public function getTitle() {
        return $this->title;
    }

    /**
     * @param mixed $title
     */
    public function setTitle($title) {
        $this->title = $title;
    }

    /**
     * @return mixed
     */
    public function getMessage() {
        return $this->message;
    }

    /**
     * @param mixed $message
     */
    public function setMessage($message) {
        $this->message = $message;
    }

    /**
     * @return false|string
     */
    public function getDate() {
        return $this->date;
    }

    /**
     * @param false|string $date
     */
    public function setDate($date) {
        $this->date = $date;
    }

    /**
     * @return mixed
     */
    public function getData() {
        return $this->data;
    }

    /**
     * @param mixed $data
     */
    public function setData($data) {
        $this->data = $data;
    }

    /**
     * @return false|string
     */
    public function toJSON() {
        $object[self::STATUS_TAG] = $this->status;
        $object[self::TITLE_TAG] = $this->title;
        $object[self::MESSAGE_TAG] = $this->message;
        $object[self::DATETIME_TAG] = $this->date;
        if ($this->data != null) {
            $object[self::DATA_TAG] = $this->data;
        }
        return json_encode($object);
    }

}

class SuccessResponse extends Response {

    public function __construct($title, $message, $data) {
        parent::__construct(ResponseStatus::OK, $title, $message, $data);
    }

}

class ErrorResponse extends Response {

    public function __construct($title, $message, $data) {
        parent::__construct(ResponseStatus::ERROR, $title, $message, $data);
    }

}

class InvalidParameterResponse extends ErrorResponse {

    public function __construct($paramName, $data) {
        parent::__construct("Invalid parameter", "Invalid parameter '" . $paramName . "'.", $data);
    }

}

class MissingParameterResponse extends ErrorResponse {

    public function __construct($paramName, $data) {
        parent::__construct("Missing parameter", "Parameter '" . $paramName . "' is missing.", $data);
    }

}

class SecurityResponse extends ErrorResponse {

    public function __construct($data) {
        parent::__construct("Access denied", "Access to this service was denied. Please log in or use a properly authorized account.", $data);
    }

}

class UnknownFailureResponse extends ErrorResponse {

    public function __construct($message, $data) {
        parent::__construct("Unknown failure", $message, $data);
    }

}

File: java/src/com/panickapps/response/Response.java
package com.panickapps.response;

import com.google.gson.JsonObject;

import java.util.Date;

/**
 * Models an HTTP response and its attributes.
 */
public class Response implements JsonSerializable {

    private static final String STATUS_TAG = "status";
    private static final String TITLE_TAG = "title";
    private static final String MESSAGE_TAG = "message";
    private static final String DATETIME_TAG = "datetime";
    private static final String DATA_TAG = "data";

    private ResponseStatus status;
    private String title;
    private String message;
    private Date date;
    private JsonObject data;

    /**
     * Creates a Response object.
     * @param status The status of the response (ok/error)
     * @param title The title of the response
     * @param message The message sent by the response
     * @param data The data requested by the client/sent by the response.
     */
    public Response(ResponseStatus status, String title, String message, JsonObject data) {
        this.status = status;
        this.title = title;
        this.message = message;
        date = new Date();
        this.data = data;
    }

    /**
     * Creates a Response object.
     * @param status The status of the response (ok/error).
     * @param title The title of the response.
     * @param message The message sent by the response.
     */
    public Response(ResponseStatus status, String title, String message) {
        this(status, title, message, null);
    }

    /**
     * Retrieves the status of the response.
     * @return (ok or error) ResponseStatus
     */
    public ResponseStatus getStatus() {
        return status;
    }

    /**
     * Retrieves the title of the response.
     * @return title (String).
     */
    public String getTitle() {
        return title;
    }

    /**
     * Retrieves the message of the response.
     * @return message (String).
     */
    public String getMessage() {
        return message;
    }

    /**
     * Retrieves the date-time of the response.
     * @return Date.
     */
    public Date getDate() {
        return date;
    }

    /**
     * Retrieves the data of the response.
     * @return JsonObject.
     */
    public JsonObject getData() {
        return data;
    }

    /**
     * Sets the status of the response.
     * @param status The status of the response.
     */
    public void setStatus(ResponseStatus status) {
        this.status = status;
    }

    /**
     * Sets the title of the response.
     * @param title The title of the response.
     */
    public void setTitle(String title) {
        this.title = title;
    }

    /**
     * Sets the message of the response.
     * @param message The message of the response.
     */
    public void setMessage(String message) {
        this.message = message;
    }

    /**
     * Sets the data of the response.
     * @param data The data of the response.
     */
    public void setData(JsonObject data) {
        this.data = data;
    }

    public String toJSON() {
        JsonObject responseJson = new JsonObject();
        responseJson.addProperty(STATUS_TAG, status.toString());
        responseJson.addProperty(TITLE_TAG, title);
        responseJson.addProperty(MESSAGE_TAG, message);
        responseJson.addProperty(DATETIME_TAG, ResponseDateTimeFormat.STANDARD.format(date));
        if (data != null) {
            responseJson.add(DATA_TAG, data);
        }
        return responseJson.toString();
    }

    @Override
    public String toString() {
        return toJSON();
    }

}
